paramiko/file.py

```diff
@@ -372,7 +372,10 @@ def write(self, data):
 
         :param str/bytes data: data to write
         """
-        data = b(data)
+        if isinstance(data, text_type):
+            # GZ 2017-05-25: Accepting text on a binary stream unconditionally
+            # cooercing to utf-8 seems questionable, but compatibility reasons?
```

Comment:
There's an important distinction from `b()` in this change, in that it allows all other types through. This *is* a tradeoff, callers get duck typing, but internal library interfaces can't rely on having a bytes type. For `BufferedFile` we get sane type handing when writing to `BytesIO` later so just passing along is best.

The reason I added this comment is that while handling here makes sense on existing branches for compatibility reasons, I'm not sure it does for master. It might be better to just raise if given unicode, and add and encode step/encoding parameter to higher level interfaces.